Remove developer __tostring method as functionality integration in game and game-modes in favor of new ToString function.

Acecool opened the pull request:
This pull request removes __tostring integration for functionality throughout the game ( where tostring( color ), tostring( vector ), or tostring( angle ) is used, they're changed to color:ToString( ), angle:ToString( ), vector:ToString( ) ) so that data-types with identical __tostring output ( such as angle and vector ) can be altered to something legible such as Vector( x, y, z ) and Angle( p, y, r ) with the arguments replaced with %d in string.format by developers without breaking gamemodes such as sandbox and DarkRP or the game which uses __tostring to explode "x y r" or "p y r" as arguments, or whatever...

This effectively allows devs to control __tostring without the game relying on that output ( because __tostring is a developer function to read the data, not for serialization )

I also removed a few empty line tabs / spaces in one or more files.
I also removed, in a few files, tostring localization which was never used in said files

Lexicality commented on garrysmod/lua/includes/modules/constraint.lua:
```diff
@@ -340,30 +340,31 @@ function CreateKeyframeRope( Pos, width, material, Constraint, Ent1, LPos1, Bone
-			rope:SetKeyValue( k, tostring( v ) )
-	
+
+			local value = ( isangle( v ) || isvector( v ) || ( istable( v ) && v.ToString ) ) && v:ToString() || tostring( v )
```
You're passing keyvalues to the engine which accepts vectors etc as `X Y Z`, not `Angle(X, Y, Z)`.
Everything in this file should be using native tostring.

Acecool replied:
This updates this line to use the :ToString( ) method which returns "x y z" or "p y r" or "r g b a" so that it no longer relies on tostring( ) meaning developers are free to alter __tostring meta functions to give clean output such as Vector( x, y, z ) for data-types that have identical formats ( Angle and Vector, for instance ).

Lexicality commented on garrysmod/lua/includes/util/color.lua:
```diff
@@ -36,16 +36,27 @@ function IsColor( obj )
 --[[---------------------------------------------------------
+	Returns color as a string
+-----------------------------------------------------------]]
+function COLOR:__tostring()
```
`COLOR.__tostring = COLOR.ToString`

Lexicality replied:
(Actually nevermind - you'll want to leave `__tostring` to return only the numbers if you adapt `ToString` to output `Colour(R, G, B, A)` yes?)

Acecool replied:
I want to allow developers to alter __tostring to output something like Color( r, g, b, a ) or Colour, but ToString lets the serialized data be used in key-value, etc...

Falco suggested a better name, instead of ToString something along the lines of Show and I am swinging that route... ToString seems too vague and too related to __tostring or tostring( x ) so Show( ) or Serialize( ) or some other name may be better suited and I don't mind altering it for this update ( because it paves the way for all of the __tostrings to be created / modified ).

Lexicality commented on garrysmod/lua/vgui/dcolormixer.lua:
```diff
@@ -307,7 +307,7 @@ function PANEL:UpdateConVar( strName, strKey, color )
 	local col = color[strKey]
 
-	RunConsoleCommand( strName, tostring( col ) )
+	RunConsoleCommand( strName, col:ToString() )
```
If `COLOR:ToString` changes output - this will break everything that uses DColorMixers.

Acecool replied:
Yes, an alternative / better name would be better suited as ToString seems to confuse rather than say what it should be such as Show / Serialize / etc.... I was going to use ToSafeString but went with ToString but I am thinking of renaming this before this is accepted...

Acecool commented:
NOTE: ToString ( Was originally going to be ToSafeString ) is too vague for what this is....

Falco suggested the name Show( ) and I'm leaning the way of renaming it something along that line or Serialize( ) or something else so that __tostring, tostring( x ) aren't confused with this.... The whole purpose of this is so that __tostring can be changed to anything the dev wants ( for debugging output using print statements ) without breaking the game, or game-modes which rely on an easy method to create a space-delimited output of these data-types...

notcake commented:
no

meepdarknessmeep commented:
no

LennyPenny commented:
no

notnanocat commented:
no

LennyPenny commented:
On a serious note, how is __tostring a "developer" function?

robotboy655 commented:
I can see what you are trying to do here, but if you, or anyone else overrides __tostring, it is honestly their own problem. Same goes for using tostring() to save values of vectors, etc to files or w/e.

This change pretty much only benefits one person - you. I bet you can understand why this is not worth merging.

Acecool commented:
This benefits more than me Robot, and I have no idea what you have against me... print is a developer function because you use print statements to debug data, etc... There are other functions to output data to the client in game, not in console...

__tostring is supposed to output data in a way that lets us know what the data is, and the values... Vector and Angle BOTH use "0 0 0" as the format, and Color uses "0 0 0 0".. It is bad development practices to use tostring to serialize data.

This corrects that BAD decision and allows __tostring to be changed so the data output by data-types can be anything that you want without breaking the game or game-modes. If you really think this is a bad Idea then you need to go to school and learn programming 101.

Re-Open this.

Zaddion commented:
I don't see why this shouldn't be added. To be honest, it does make since and I could see more than just Acecool using this,

thegrb93 commented:
Without __tostring the way it is, I can't do print(myvector) which would suck.

Acecool commented:
This doesn't alter any __tostrings in this pull, meaning you can still use print( myvector ) and it'll still output 0 0 0...

What this does is unlink __tostring from data-serialization the game uses so that __tostring can be changed to output something such as "Vector( 0, 0, 0 )", and "Angle( 0, 0, 0 )" instead of potentially confusing data-types because both Vector and Angle use the exact same output format of "0 0 0"

thegrb93 commented:
I know, I was arguing your point about __tostring

waut3r commented:
This adds functionality, so I don't see what's wrong with it.

meepdarknessmeep commented:
what's the point of having a ToString() metafunction if we already have an automatic tostring metamethod that works on everything? I just see it as bloated code we will never need.

neico commented:
http://www.lua.org/manual/5.1/manual.html#pdf-file:write
> To write other values, use tostring or string.format before write.

tostring calls __tostring if set, so it seems perfectly fine to me to use that for serialisation, no one cares if that function outputs `Angle( 0, 0, 0 )` or `0 0 0`

What counts is that the format is consistent across the entire vanilla code.
you can still add a function like this if you really need to:

```lua
local meta = FindMetaTable( "Angle" )

-- This could actually be prototyped by storing it as a local function and then assigning it to multiple metatables~
function meta:ToString()
	return meta.MetaName .. "( " .. self.p .. ", " .. self.y .. ", " .. self.r .. " )";
end

-- recommended way to only use it on your local code to avoid bloats like the one above...
-- actually one could make it a string.Angle() function as well,
-- which can be added to vanilla without many complaining, but it's still over-complicating things
local function AngleToString( pAngle )
	return type( pAngle ) .. "( " .. pAngle.p .. ", " .. pAngle.y .. ", " .. pAngle.r .. " )"
end
```

There are more efficient ways to do this but I'm not trying to write actual production ready code here so...

Acecool commented:
I renamed ToString to Serialize and added 2 optional arguments ( delimiter and postfix ) to make more sense of what the function aims to accomplish.

print( ) is a developer function because developers use it to output data, use it to output data-type data, use it for debugging, etc... If a developer tries printing angle_zero:Forward( ) as an example, it'll output "0 0 0" and the developer may believe that it is still an Angle despite Forward converting the Angle into a normalized Vector.

The goal of this pull is to unlink __tostring from serialization in the game so __tostring can be altered without breaking the game. Why should the print function need to be rewritten to look for ToString just so some data can be serialized in some other place using the output? This is a practical solution to a poor design choice in code... You can Serialize the data without arguments and the delimiter will be a space and the postfix will be nothing by default. You can add postfix to separate colors, vectors, or angles if you want to serialize them in one large string without needing to use ..";" at the end of the tostring by using: col:Serialize( nil, ";" ), or if you want to change the space you can using the first argument...

I have no idea where this wouldn't make sense to merge this... The whole point of coding is to create libraries, etc.. which will work with other elements. By hardcoding in the meta-function __tostring as a serialize ( into some format ) doesn't make sense to me when data-types can be confused with each-other because of identical outputs, and print isn't normally used outside of debugging / development work so, again, this makes sense to unlink it so print can be used with clean data-type output without breaking the game or game-modes.

I corrected this same issue in DarkRP and Falco thought it was a good idea because he was also thinking of rewriting __tostrings which would go well with his simplerr / clean-error-output-system...

Acecool commented:
Because this is closed, the latest changes aren't showing so I'll need to repost to show latest changes.

robotboy655 commented:
As I said, if you override default methods, it is your own fault.

neico commented:
> DarkRP

I think I've seen enough to think the rest for myself...

Acecool commented:
Updated pull: https://github.com/garrynewman/garrysmod/pull/920

Robot, I have no idea what you have against me ( even going so far as deleting older pull requests of mine just so you could re-post them ), but it needs to stop. This is an obvious solution to a very real problem. By not merging it, you're allowing a bad-design choice to stay in the game which can cause the game to break.

Why should we need to rewrite the default functionality of Lua, Lua meta-tables and Lua meta-table functions or print to have proper output because someone was too lazy to properly serialize the data, or create a helper-function to do it for them?

I've done it, this is an obvious solution to merge.

robotboy655 commented:
I have nothing against you, and I am pretty sure that one your PR that I closed and included in mine is not "closing many of your PRs and recreating them".

I just looked through every PR you ever suggested, I never closed ANY of them. I merged quite a few though. So I don't know why you need to exaggerate or lie about my actions towards you.

Acecool commented:
I apologize; to me it appeared as though there was an issue with me but if it isn't then, again, please accept my apologies. I won't mention that again.
